### PR TITLE
[wgpu-hal] Upgrade to `ash 0.38`

### PR DESCRIPTION
**Pull request by MarijnS95**
https://github.com/ash-rs/ash/releases/tag/0.38.0

In this release a lot of breaking changes have been made: builder structs are dropped in favour of always having a lifetime parameter available on every raw Vulkan structure (when they contain one or more pointers).  This massively contributes to lifetime (and mutable aliasing) safety, but requires significant changes to some uses in `wgpu-hal`.

All function pointer loaders for both Vulkan core and extensions have moved into the root of `ash::`, making `ash::vk::` a more pure `sys`-like module.  Extensions have their own `ash::<prefix>::<ext>` module to clearly separate and group their items from the core.  Besides `NAME` and the usual `*Fn` wrappers, the handwritten `extensions` module is now only available via this path.  This to combat the previous inconsistency between `ash::KhrSomeExtFn::name()` vs `ash::extensions::khr::SomeExt::name()`.

The Vulkan core clearly splits functions across `device` and `instance` functions, to keep functions that can be loaded without a dispatch-table for a device via `get_device_proc_addr()` apart from instance functions. This concept has now been applied to extension functions, making it possible to load them optimized for a device too (when the shared table would previously include instance functions and require the whole thing to be loaded via `get_instance_proc_addr()`), and in the rare case of 3 hand-written extension wrappers: load instance functions via `get_device_proc_addr()` resulting in `NULL` pointers.

Finally, a few new helpers like `_as_slice()` and `_as_c_str()` are available (the former only for statically-sized struct-owned arrays with a length delimiter field) to simplify commonly written patterns.

**Connections**
_Link to the issues addressed by this PR, or dependent PRs in other repositories_
https://github.com/gfx-rs/wgpu/pull/5501

**Description**
_Describe what problem this is solving, and how it's solved._
New crate release of `ash` that users will likely want to start using, especially since it exposes a many more Vulkan extensions (generated against a newer spec) and has significant lifetime-safety improvements and other usability-refactors.

**Testing**
_Explain how this change is tested._
`cargo r -p wgpu-examples` for a few of them.

**Checklist**

- [x] Run `cargo fmt`.
- [x] Run `cargo clippy`. If applicable, add:
  - [ ] `--target wasm32-unknown-unknown`
  - [ ] `--target wasm32-unknown-emscripten`
- [ ] Run `cargo xtask test` to run tests.
  This fails on the `water` test, but also on `trunk`. As seen when running `cargo r -p wgpu-examples water`, a lot of `SYNC-HAZARD-WRITE-AFTER-WRITE` validation issues are detected.
- [x] Add change to `CHANGELOG.md`. See simple instructions inside file.

---

I realize that this is a bit of a sloppy PR with some `TODO`s, which I'd like to discuss to find the cleanest way forward :)

### CODE REVIEW

File: wgpu-hal/src/vulkan/device.rs
Comment (MarijnS95) [inline wgpu-hal/src/vulkan/device.rs]: Casual bug-fix here: `write.descriptor_count` should be equal to `acceleration_structure_count` in `vk::WriteDescriptorSetAccelerationStructureKHR`. Unless the selected range of `desc.acceleration_structures` (is `entry.count`) is always 1?

File: wgpu-hal/src/vulkan/instance.rs
Comment (MarijnS95) [inline wgpu-hal/src/vulkan/instance.rs]: Any reason to call the Windows function when the `raw-window-handle` contains `hinstance` already? Probably and oversight since, as history suggests, when this code was introduced in 2021 on `raw-window-handle` the field was already there.

File: Cargo.toml
Comment (teoxoy) [inline Cargo.toml]: They are only used for some of the crates due to a bug in cargo. The bug has been fixed but we didn't yet re-enable full workspace inheritance; see https://github.com/gfx-rs/wgpu/issues/3407.

File: wgpu-hal/src/vulkan/instance.rs
Comment (teoxoy) [reply wgpu-hal/src/vulkan/instance.rs]: I think it was originally introduced in https://github.com/gfx-rs/gfx/pull/1653 and since then copy-pasted.

But given it's optional, could it be that it's not set? I see `winit` sets it but it doesn't seem to be a requirement of `raw-window-handle`.

File: wgpu-hal/src/vulkan/command.rs
Comment (teoxoy) [inline wgpu-hal/src/vulkan/command.rs]: I think we should inline it.

File: wgpu-hal/src/vulkan/adapter.rs
Comment (teoxoy) [inline wgpu-hal/src/vulkan/adapter.rs]: `enabled_extensions` are device extensions. I think we can consult `self.instance.extensions` for instance extensions.

File: wgpu-hal/src/vulkan/device.rs
Comment (teoxoy) [inline wgpu-hal/src/vulkan/device.rs]: I think we should create a new data structure based on a `Box<[MaybeUninit<T>]>`. It would only need an extend method that returns a `&[T]`.

**Review by teoxoy — CHANGES_REQUESTED** [review]

File: wgpu-hal/src/vulkan/device.rs
Comment (MarijnS95) [reply wgpu-hal/src/vulkan/device.rs]: Sure, I've thought of that in the beginning, but the nature of the borrow checker requires the storage (a `Vec` or `Box`) to live separately from the mechanism that's mutably borrowed _or moved_ to keep track of the remaining space of the slice to write. `split_first_mut()` and `split_at_mut()` are critical helpers here as they splice the lifetime of the underlying `Vec`.

Implementing a data structure like you suggest, with a signature like:

```rust
fn extend(&mut self, values: IntoIterator<Item = T>) -> &[T] {}
```

is unfortunately impossible as it makes the returned slice _which `ash`'s builder-pattern correctly borrows for the remaining duration of `writes: Vec<vk::WriteDescriptorSet<'here>>`_ also borrow on `self`, which cannot be reused anymore in the next iteration of the loop. That is also why this implementation is repeatedly splitting the slice and reassigning the remainder to neatly get the lifetimes correct _without `unsafe`_.

I've pushed a wrapper now which re-adds `MaybeUninit` behaviour, relying on `Vec::with_capacity(x).spare_capacity_mut()` to get access to a `&mut [MaybeUninit<T>]` unfortunately a similar initializer for `box` is still unstable:

https://doc.rust-lang.org/std/boxed/struct.Box.html#method.new_uninit_slice

Let me know if there are any more improvements to this system!

File: wgpu-hal/src/vulkan/instance.rs
Comment (MarijnS95) [reply wgpu-hal/src/vulkan/instance.rs]: Yeah, I'm not actually certain why and in what case it is supposed or allowed to be left at zero. Nor if it makes sense for `wgpu` to fall back to `GetModuleHandleW()` if it's unset.

I've never seen it being `None`/`0`, at least not with `winit`.

[Vulkan spec demands it to be valid](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkWin32SurfaceCreateInfoKHR.html), and I've converted the unwrap into returning a proper `InstanceError`: should we do the same to the `expect()`s at `Xlib` and `Xcb`?

File: wgpu-hal/src/vulkan/adapter.rs
Comment (MarijnS95) [reply wgpu-hal/src/vulkan/adapter.rs]: Correct, this is an instance extension that should not be enabled on the `instance`, I just didn't realize that the list of extensions was this close in reach :)

File: wgpu-hal/src/vulkan/instance.rs
Comment (teoxoy) [reply wgpu-hal/src/vulkan/instance.rs]: I'm not sure, are these errors recoverable? I would generally lean towards errors for libraries; although in this case `InstanceError` is opaque... 😅

Comment (teoxoy) [reply wgpu-hal/src/vulkan/instance.rs]: I think we should keep things consistent though.

@cwfitzgerald do you have an opinion on panic vs error for these?

Comment (MarijnS95) [reply wgpu-hal/src/vulkan/instance.rs]: Not sure either, I'm not a windows expert on this front. We could keep the `.expect()` and ask the `raw-window-handle` devs what made them consider this optional?

Seems like there are multiple ways to get this opaque handle anyway, via `GetModuleHandleW(NULL)` for the current module or [`GetWindowLongPtrA(hwnd, GWLP_HINSTANCE)`](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-getwindowlongptra)...

File: wgpu-hal/src/vulkan/device.rs
Comment (teoxoy) [reply wgpu-hal/src/vulkan/device.rs]: Right, this is a problematic case for rust's borrow checker.

I think the new data structure is still better than nothing, thanks for adding it!

Comment (MarijnS95) [reply wgpu-hal/src/vulkan/device.rs]: Correct, the pattern is tricky for the borrow checker but _at least_ thanks to functions like `split_at_mut()`/`split_first_mut()` we can make them work in a lifetime+uniqueness-checked way while only doing one pre-allocation.

Would still be better if some of the `MaybeUninit` helpers are stabilized, but so be it :)

File: wgpu-hal/src/vulkan/adapter.rs
Comment (teoxoy) [inline wgpu-hal/src/vulkan/adapter.rs]: I don't think we should panic here.

File: wgpu-hal/src/vulkan/device.rs
Comment (teoxoy) [inline wgpu-hal/src/vulkan/device.rs]: I don't think this needs to be pinned. `CString` implements `Unpin`.

The pointer returned by `CString.as_ptr()` will be valid and won't change as long as we don't drop the `CString`.

File: wgpu-hal/src/vulkan/device.rs
Comment (teoxoy) [inline wgpu-hal/src/vulkan/device.rs:1442]: Could you add a doc comment explaining a bit why we need this? For future reference.

File: wgpu-hal/src/vulkan/device.rs
Comment (teoxoy) [inline wgpu-hal/src/vulkan/device.rs]: Do we need to pin something?

File: wgpu-hal/src/vulkan/device.rs
Comment (teoxoy) [inline wgpu-hal/src/vulkan/device.rs]:
```suggestion
        let mut acceleration_structure_infos = Vec::with_capacity(desc.acceleration_structures.len());
```

File: wgpu-hal/src/vulkan/device.rs
Comment (teoxoy) [inline wgpu-hal/src/vulkan/device.rs]:
```suggestion
        let mut raw_acceleration_structures = Vec::with_capacity(desc.acceleration_structures.len());
```

File: wgpu-hal/src/vulkan/instance.rs
Comment (teoxoy) [inline wgpu-hal/src/vulkan/instance.rs:206]: What are the safety requirements here?

File: wgpu-hal/src/vulkan/instance.rs
Comment (teoxoy) [inline wgpu-hal/src/vulkan/instance.rs:572]: What are the safety requirements here?

File: wgpu-hal/src/vulkan/instance.rs
Comment (teoxoy) [inline wgpu-hal/src/vulkan/instance.rs:613]: and here?

File: wgpu-hal/src/vulkan/adapter.rs
Comment (teoxoy) [inline wgpu-hal/src/vulkan/adapter.rs]: Can we rename this to something else? `properties2`?

File: wgpu-hal/src/vulkan/adapter.rs
Comment (teoxoy) [inline wgpu-hal/src/vulkan/adapter.rs]: same here, `features2`?

**Review by teoxoy — CHANGES_REQUESTED** [review]

File: wgpu-hal/src/vulkan/device.rs
Comment (teoxoy) [reply wgpu-hal/src/vulkan/device.rs]: Makes sense to me.

@daniel-keitel do you have more context from https://github.com/gfx-rs/wgpu/pull/3507?

File: wgpu-hal/src/vulkan/device.rs
Comment (MarijnS95) [reply wgpu-hal/src/vulkan/device.rs]: Why is why `Unpin` has never made sense to me, since `(C)String` contain a pointer internally. If I can perform destructive actions on the contents of `(C)String` (e.g. `mem::replace()`, or appending to the string causing a realloc), how can I guarantee the memory address remains pinned?

File: wgpu-hal/src/vulkan/device.rs
Comment (MarijnS95) [reply wgpu-hal/src/vulkan/device.rs]: If my understanding of `Pin` is correct (see above: **it is not**) it _should_ prevent us from causing destructive actions on the pointer (`.push()`, `.resize()`, `.clear()`) after having `let mut` access to a `Vec`, while we can still `Deref` it into a `&mut [T]` to modify the elements without affecting where they are allocated.

I don't think that deref mechanism works for `spare_capacity_mut()` though.

(And having mutable borrows via `ExtendStack` prevents us from modifying the `Vec` anyway)

Please correct me about this.

File: wgpu-hal/src/vulkan/instance.rs
Comment (MarijnS95) [reply wgpu-hal/src/vulkan/instance.rs:572]: Same as any other Vulkan call through `ash`. They just didn't happen to be marked `unsafe` before.

File: wgpu-hal/src/vulkan/device.rs
Comment (teoxoy) [reply wgpu-hal/src/vulkan/device.rs]: I looked at the docs and there is no way to mutate `CString`'s content. You can only do `mem::replace()` on the `CString` itself so the only way the pointer returned by `.as_ptr()` becomes invalid is if you drop the `CString`.

I find rust's safety docs to be top-notch and the docs of `.as_ptr()` indicate that the only invariant you have to care about is keeping the `CString` alive.

File: wgpu-hal/src/vulkan/device.rs
Comment (teoxoy) [reply wgpu-hal/src/vulkan/device.rs]: As long as there are references to subslices of the `Vec` around, we can't call any of `Vec`'s methods that take `&mut Self`, so we should be good.

File: wgpu-hal/src/vulkan/device.rs
Comment (MarijnS95) [reply wgpu-hal/src/vulkan/device.rs]: You are right that `CString` makes this easy. In the broader picture of `Pin` and [`Unpin`, its example](https://doc.rust-lang.org/std/marker/trait.Unpin.html) suggests a `String` that seems to be mutated in place and voids its pointer placement? `String::as_ptr()` doesn't have any documentation like `CString`.

Comment (MarijnS95) [reply wgpu-hal/src/vulkan/device.rs]: Actually, never mind. In that example the `Pin` doesn't own the `String` but a borrow of the `String`, which is different. I think my understanding and interpretation of `Pin` still holds then. After all, `String` is safe to move around, the thing it points to internally is not.

**Comment by cwfitzgerald** [comment]
Really appreciate you spearheading the work on upgrading! Some conflicts need to be resolved and want Teo to give it one last look, but then we should be g2g.

**Review by teoxoy — COMMENTED** [review]
There are 2 remaining discussions that we should address
- [Should some missing params in surface creation be errors or panics?](https://github.com/gfx-rs/wgpu/pull/5504#discussion_r1555435083) cc @cwfitzgerald @Wumpf 
- [Removing the Pin around CString](https://github.com/gfx-rs/wgpu/pull/5504#discussion_r1559382929)

**Comment by ErichDonGubler** [comment]
@MarijnS95: Do you think you'll be able to attend to this soon? I'm excited to see quite a few duplicate dependencies like `libloading` 0.7 disappear once this lands. 🙂 I filed #5649 in anticipation. No pressure, though!

**Comment by MarijnS95** [comment]
Before continuing further, it has to be said that those repeated "Changes requested" are quite off-putting. In the projects I work on that button is reserved as a heavy ban-hammer that basically means a full "NAK, go back to the drawing board". Any other nits and smaller comments use just that: the "leave a comment" mode which is neither an approval nor a full-on rejection.

Especially if most review comments are merely answering my questions.

---

Looks like we only have the surface creation panics to discuss, and the occasional acceleration structure descriptor-set-write drive-by-bugfix.

**Comment by cwfitzgerald** [comment]
> Before continuing further, it has to be said that those repeated "Changes requested" are quite off-putting. In the projects I work on that button is reserved as a heavy ban-hammer that basically means a full "NAK, go back to the drawing board". Any other nits and smaller comments use just that: the "leave a comment" mode which is neither an approval nor a full-on rejection.

Definitely don't intend them to be off-putting! We use it basically as the middle of three options: reject entirely, need modifications, and approve. The "reject entirely" option for us would be commenting our disapproval of the strategy, then closing the PR entirely. Requesting changes also kicks the PR off our team review queue until review is re-requested. I think we could do better at communicating that the expected flow is `PR -> request changes -> fix/discuss -> re-request review -> (repeat) -> approval -> merge`.

**Comment by MarijnS95** [comment]
> Requesting changes also kicks the PR off our team review queue

That's something I haven't considered yet, mostly being a lone-wolf maintainer where the TODO backlog is always "somehow further all PRs that are still open".

**Comment by jimblandy** [comment]
> Before continuing further, it has to be said that those repeated "Changes requested" are quite off-putting. In the projects I work on that button is reserved as a heavy ban-hammer that basically means a full "NAK, go back to the drawing board".

I certainly don't intend my "changes requested" to be discouraging to anyone! I've always took them to mean exactly what they say: I'm requesting changes. I think the distinction I'm hoping to make is: If I only have suggestions, then I'll note them in the comments and "Approve", but if I see a bug or something that I think needs to be done differently, then I mark it "Changes requested". If I really wanted to end discussion of the whole approach, I'd close the PR. (What Connor said, basically.)

Anyway - just to say it out loud: @MarijnS95, thank you very much for this PR! Having `wgpu` be an open-source project is such a welcome contrast to most of the Firefox code base, whose contributors are almost exclusively Mozilla employees. I don't think Firefox's WebGPU implementation could possibly succeed with such a small team (me, @teoxoy, @ErichDonGubler, and @nical part-time) if we didn't also have contributions coming in on the wgpu side.

edit: [this comment](https://github.com/gfx-rs/wgpu/pull/5504#issuecomment-2091294413) from Erich is also a reflection of the value Mozilla, for our sake, finds in these kinds of contributions.

**Comment by ErichDonGubler** [comment]
I think our community chat would be a good place to continue this discussion about perceptions with PR statuses, so I started a thread in Matrix: https://matrix.to/#/!FZyQrssSlHEZqrYcOb:matrix.org/$_lc2jnjTVMqi3SD2nsjSYmFdSSAlt7cO0wBahOcfT-0?via=mozilla.org&via=matrix.org&via=envs.net

File: wgpu-hal/src/vulkan/instance.rs
Comment (cwfitzgerald) [reply wgpu-hal/src/vulkan/instance.rs]: InstanceError is fine here I think.

**Comment by cwfitzgerald** [comment]
I think once merge conflicts are resolved, it is ready to go

**Review by teoxoy — APPROVED** [review]
Looks good to me as well!

File: wgpu-hal/src/vulkan/command.rs
Comment (MarijnS95) [inline wgpu-hal/src/vulkan/command.rs:457]: :disappointed:

**Review by cwfitzgerald — APPROVED** [review]
CI is mad, but lets go :)

**Comment by MarijnS95** [comment]
@cwfitzgerald Maybe I broke that with the change in features for `winapi`? Though the `libloaderapi` feature that I moved around doesn't seem to impact whether `errhandlingapi` is enabled.

It is also a tad unfortunate that local `cargo clippy` on Rust 1.78 shows a bunch of `unnecessary qualification` warnings despite #5651 being merged, in code that wasn't touched in this PR.

**Comment by MarijnS95** [comment]
The cause of this error is now obvious. With this PR there's no direct dependency on `libloading 0.7` anymore (there is via `glutin`, but it doesn't seem to be enabled by default, or at least not for the default target(s)/crate(s)), which used to turn on `errhandlingapi` on `winapi`:

```console
$ cargo tree -e features --target x86_64-pc-windows-msvc
...
│   │   ├── wgpu-hal v0.20.0 (/home/marijn/Code/TraverseResearch/wgpu/wgpu-hal)
...
│   │   │   ├── ash feature "default"
│   │   │   │   ├── ash v0.37.3+1.3.251
│   │   │   │   │   └── libloading feature "default"
│   │   │   │   │       └── libloading v0.7.4
│   │   │   │   │           ├── winapi feature "default" (*)
│   │   │   │   │           ├── winapi feature "errhandlingapi"
│   │   │   │   │           │   └── winapi v0.3.9
...
```

So we now have one more explicit feature dependency to list in `Cargo.toml`!

**Comment by cwfitzgerald** [comment]
Thanks for the work on this and seeing this through to the end!

**Comment by MarijnS95** [comment]
No worries, I hope you can all benefit/enjoy the incremental improvements in the latest `ash` release now!

**Pull request merged**
